Queues tests

bebehei opened the pull request:
---

Edit: Fixes #563

tsipinakis commented:
That's a very good catch! We hadn't had a discussion on it but previously we assigned the ID asked anyway. The notification spec doesn't specify what to do if the ID is invalid but we should preserve the previous behaviour to not break those that use a static id replacement as the old workaround for `stack_tag`.

I consider this a bug and it was probably introduced in #552.

tsipinakis commented:
I pushed a fix @bebehei

bebehei commented on test/queues.c:
```diff
@@ -529,6 +529,67 @@ TEST test_queues_update_paused(void)
         PASS();
 }
 
+TEST test_queues_update_sicker(void)
```
`s/sicker/seep/g`.

Sorry, I totally missed that this isn't even an english word :grin:

bebehei commented on test/queues.c:
```diff
@@ -355,6 +355,65 @@ TEST test_datachange_ttl(void)
+
+TEST test_queue_stacktag(void)
+{
+        const char *stacktag = "THIS IS A SUPER WIERD STACK TAG";
```
Intended ;-)

bebehei commented on test/queues.h:
```diff
@@ -0,0 +1,97 @@
+#define ACT_CLS_ID(id) queues_notification_close_id(id, REASON_UNDEF)
+#define ACT_RPL(n) queues_notification_replace_id(n)
+#define ACT_RUN(data) queues_update(data)
+#define ACT_REDISPLAY(data) queues_upodate(data)
```
Full ACK. I first started the tests, I basically packed everything into a macro. In the meantime, I've learnt, that `ASSERTm` does a pretty awesome job in most cases.

But I hadn't been able to remove all the unused (and unnecessary) macros.

bebehei commented:
Most of the stuff is now fixed.

> I also found a bug: After a notification is shown and closed the dunst window isn't hidden, a small black rectangle keeps being displayed (and moves around the screens with the cursor as would be expected). Bisect showed 052d9f7 as the broken commit.

This is now fixed. The `visible` field of the `dunst_status` structure was absolutely redundant. I tripped too much over the word `visible`, which had different meanings in different contexts.

But shame on me, that I requested your review, but did not even bother to run dunst :face_with_head_bandage:

tsipinakis commented on src/queues.c:
```diff
@@ -121,6 +123,36 @@ static bool queues_notification_is_ready(const struct notification *n, struct du
+
+        /* don't timeout when user is idle */
+        if (status.idle && !status.fullscreen && !n->transient) {
+                n->start = time_monotonic_now();
```
Why are you modifying `n->start` in a query function? What's the point of this if block entirely?

bebehei replied:
You're never considered idle, if you're in fullscreen. Current master:

https://github.com/dunst-project/dunst/blob/ed341dfca30ed496bb056a00251af6a17d1f7a07/src/queues.c#L346

I think, I should use exactly this line in there also in the new function.

tsipinakis commented on src/queues.c:
```diff
@@ -343,68 +375,32 @@ void queues_history_push_all(void)
 
 /* see queues.h */
-void queues_check_timeouts(struct dunst_status status)
+void queues_update(struct dunst_status status)
```
These are 2 huge and complicated functions, let's break them down rather than merge them.

bebehei replied:
I don't see any benefit of splitting this method.

There are 4 parts, which are required to be executed in exactly this order and the logics involved is nowhere else necessary.

Additonally, the complex testing logic is already splitted out into `queues_notification_is_(ready|finished)`.

bebehei commented on src/notification.c:
```diff
@@ -450,6 +450,7 @@ static void notification_format_message(struct notification *n)
                 case '\0':
                         LOG_W("format_string has trailing %% character. "
                               "To escape it use %%%%.");
+                        substr++;
```
If I set `substr` to `NULL`, the `substr = strchr(substr, '%')` expression, which will get executed at the end of the loop will segfault.

bebehei commented on src/queues.c:
```diff
@@ -455,7 +455,7 @@ void queues_update(struct dunst_status status)
                        && (i_displayed = g_queue_peek_tail_link(displayed))) {
 
-                        while (i_waiting && ! queues_notification_is_ready(i_waiting->data, status, true)) {
+                        while (i_waiting && ! queues_notification_is_ready(i_waiting->data, status, false)) {
```
> From here, the bug and the solution becomes obvious: We can simplify the entire function to
> 
> ```
> return n->fullscreen == FS_SHOW || !fullscreen;
> ```

I'd love to simplify it as that. But it's exactly that corner case you're pointing out. Let's rephrase it in language:

fullscreen, visible = (1,1) means that the notification got inserted and shown while in a "normal" mode, but afterwards you switched a window to fullscreen.

And that's the exact point between delay and pushback:

https://github.com/dunst-project/dunst/blob/d879d70da060ea78fe735d62249a0afdf3e61bc8/dunstrc#L318-L323

bebehei commented on src/queues.c:
```diff
@@ -121,6 +123,36 @@ static bool queues_notification_is_ready(const struct notification *n, struct du
+ *
+ * @param n the notification to check
+ * @param status the current status of dunst
+ * @returns true, if the notification isn't timed out, otherwise false
```
Thanks for pointing out such a stupid nobrainer.

tsipinakis commented on src/queues.c:
```diff
@@ -134,15 +134,15 @@ int queues_notification_insert(struct notification *n)
         /* Do not insert the message if it's a command */
         if (STR_EQ("DUNST_COMMAND_PAUSE", n->summary)) {
-                pause_displayed = true;
+                dunst_status(S_RUNNING, false);
```
I don't think we'd ever need to actively signal a command line tool as it usually doesn't hang around, makes a request and exits.

But either case aside IMO it's best to avoid needless abstraction on a "might need this" basis, especially one as simple to implement later as this.

bebehei replied:
I agree with you IMO, but I'm already working on #445 and to have the ability for a signal is mandatory for #216.

Also the status field is declared static in dunst.c.

tsipinakis commented on test/queues.h:
```diff
@@ -0,0 +1,57 @@
+        struct notification *n = notification_create();
+
+        if (timeout != -1)
+                n->timeout = S2US(timeout);
```
Huh, here's a case I didn't consider: What if timeout is `-1`? It will be multiplied anyway and there are some places in the code we explicitly check for `-1` like [`rules.c`](https://github.com/dunst-project/dunst/blob/master/src/rules.c#L17), need to check if there are breakages with this.

bebehei replied:
`-1` is the default for the notification's timeout:

https://github.com/dunst-project/dunst/blob/d786381cb0d52788de56732fe0f781e4823b6e99/src/notification.c#L312

A timeout of `-1` doesn't interfere with rules at all.

Also in `notification_init` every timeout below 0 will become the default timeout.

https://github.com/dunst-project/dunst/blob/d786381cb0d52788de56732fe0f781e4823b6e99/src/notification.c#L340-L342

tsipinakis commented on test/queues.h:
```diff
@@ -0,0 +1,57 @@
+
+#define NOT_LAST(n) do {ASSERT_EQm("Notification " #n " should have been deleted.", 1, notification_refcount_get(n)); g_clear_pointer(&n, notification_unref); } while(0)
+
+static inline struct notification *test_notification(const char *name, gint64 timeout)
```
What's the point of the `timeout` parameter? In the test cases you change the field you are testing on the next instruction, why have a parameter that's `-1` most of the time?

bebehei replied:
?

tsipinakis replied:
Nevermind, I took a look at the first few test commits and noticed that `timeout` was `-1` for all of them and assumed it was a redundant parameter. A quick `grep` told me it's actually used more often.

tsipinakis commented on src/dunst.c:
```diff
@@ -24,6 +25,34 @@
+
+/* see dunst.h */
+void dunst_status(const enum dunst_status_field field,
+                  bool value)
```
`value` can also be `const`, but is there a point declaring variables on the stack as `const`?

bebehei replied:
"Just because I can":tm:

Actually there is no specific reason. I started doing it with strings (where it's a useful thing).

But yes, either both params should be `const` or not.

tsipinakis commented on test/queues.c:
```diff
@@ -166,8 +166,116 @@ TEST test_queue_teardown(void)
+
+        queues_update(STATUS_NORMAL);
+        ASSERT_IN_RANGEm("The notification has to get closed in less than its timeout",
+               S2US(10), queues_get_next_datachange(time_monotonic_now()), S2US(10));
```
`ASSERT_IN_RANGE` doesn't make sense here, for one it assumes the value is double which makes the error message confusing by using scientific notation and secondly if `S2US(15)` is returned it still passes as it checks for `+/- S2US(10)`.

bebehei replied:
Oh, well it **does** make sense. Most importantly it has to be less than the actual timeout. But also it should be non-negative (this would be the representation for not sleeping).

tsipinakis commented on test/queues.c:
```diff
@@ -166,8 +166,116 @@ TEST test_queue_teardown(void)
+        queues_notification_insert(n);
+        queues_update(STATUS_NORMAL);
+        ASSERT_IN_RANGEm("The timeout of the second notification has to get used as sleep time now.",
+               S2US(n->timeout), queues_get_next_datachange(time_monotonic_now()), S2US(n->timeout));
```
Same thing in this test.

bebehei replied:
There's something worse. Do you spot the problem?

tsipinakis replied:
Oh, calling `S2US` on `timeout` nice catch!

tsipinakis commented on test/queues.c:
```diff
@@ -2,6 +2,36 @@
+        n = test_notification("n3", 0);
+        queues_notification_insert(n);
+
+        QUEUE_LEN_ALL(1,1,1);
```
It'd be a good idea to also check as you're adding the notifications to ensure that they're being put in the correct queue rather than only at the end.

bebehei replied:
It's not important which notification is where. It's just an assertion, that every queue will contain an element.

And it seems to be redundant, but think about it like an assertion. If the `queues_notification_insert` and `queues_update` will fail, the test doesn't step into the problem, that `queues_length_*` reports the wrong size.

tsipinakis commented on test/queues.c:
```diff
@@ -2,6 +2,36 @@
+        ASSERT_EQm("Queue displayed has to contain an element",
+                   1, queues_length_displayed());
+        ASSERT_EQm("Queue history has to contain an element",
+                   1, queues_length_history());
```
These are redundant: if `QUEUE_LEN_ALL` fails above they won't get run either way, if it succeeds they will pass.

bebehei replied:
I already wanted to rebase and remove it. The test seems duplicated.

Bit it isn't: The test wants to test `queues_length_*`, so we have those `ASSERT_EQm`.

tsipinakis commented:
Heads up: I pushed a fixup for a small bug in my id assignment commit.

bebehei commented:
> Heads up: I pushed a fixup for a small bug in my id assignment commit.

Thoughts in order:

1. Yay, thanks! 
1. Wait, all your comments which got automatically marked outdated, aren't outdated anymore!?
1. Surprise surprise, all my changes are gone! :tada: 

I'm glad that I'm a GitHub beta member (you can click on the images to see the diff):

[![Force Push IMG](https://user-images.githubusercontent.com/2073458/48968691-46e05b00-eff3-11e8-865a-a1c3435bc538.png)](https://github.com/dunst-project/dunst/compare/c646aeced84a14e016a92b3e562214fe96ea362a..69aa7c4fa942adba807cdb4ea5423f2966804297)

[![Force push tsipinakis](https://user-images.githubusercontent.com/2073458/48969179-63809100-effb-11e8-863f-f0e6a6dfb4ce.png)](https://github.com/dunst-project/dunst/compare/69aa7c4fa942adba807cdb4ea5423f2966804297..811090d77a99efbd62acb1b19e12fee602c5306e)

Please fetch the upstream pull request the next time first. If you need to fetch them very easily, there's a simple refspec available to fetch the GitHub PRs automatically: https://bebehei.de/github-prs/

tsipinakis commented:
> Surprise surprise, all my changes are gone! tada

Oh crap, sorry! totally forgot you had pushed more changes in the meantime!

bebehei commented:
Well, this happens.

Would you mind giving a final review?

tsipinakis commented on test/queues.c:
```diff
@@ -364,6 +364,65 @@ TEST test_datachange_ttl(void)
+
+        n1 = test_notification("n1", 1);
+        n2 = test_notification("n2", 1);
+        n3 = test_notification("n3", 1);
```
Looking at this again the timeout parameter _does_ seem redundant. What's the point in setting it to `1` here? Any value wouldn't make a difference as we aren't advancing time.

But anyway, that's a complete nitpick everything else looks good.

bebehei replied:
I guess this is a typo introduced via vim block mode during refactoring.
